File: Data Structures & Algorithms/palindrome-partitioning/submission-0.cpp
class Solution {
private:
    vector<vector<string>>result;
public:
    bool isPal(string s){
        string tmp = s;
        reverse(s.begin(),s.end());
        return tmp == s;
    }

    void dfs(string s, int curr,int n,vector<string>path){
        if(curr >= n){
            if(count(result.begin(),result.end(),path) == 0) result.push_back(path);
            return;
        }

        for(int len = 1; len <= n; len++){
            string subString = s.substr(curr,len);
            if(!isPal(subString)){
                continue;
            }
            path.push_back(subString);
            dfs(s,curr + len, n, path);
            path.pop_back();
        }
    }

    vector<vector<string>> partition(string s) {
        //dumbass approach
        int n = s.size();
        result.clear();
        dfs(s,0,n,{});
        return result;
    }
};

File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
            n = 3
            ()

            ((()

            ()()()
            ()(())

            valid parantheses contain valid ones

            ((()))

            open/close count

        """

        result: List[str] = []

        def dfs(open_count:int, close_count: int, path: str):
            if open_count == 0 and close_count == 0:
                tmp = path
                result.append(tmp)
                return
            
            if open_count == close_count or open_count > 0:
                dfs(open_count-1, close_count, path + '(')
            

            if open_count < close_count:
                dfs(open_count, close_count-1, path + ')')
            
           
        dfs(n, n, '')
        return result
        

